Implemented close application after opening a file / folder

torbei opened the pull request:
I tried my hands on implementing issue #84 

At the moment it closes the application after opening the file (or the folder of the file) via the keyboard.

But it doesn't work when you open it via the mouse (double-clicking on it).
Someone can argue if this should be the case or not.
Personally I would want it to stay open when using the mouse because then I'm looking for something in the file and maybe opening another file after that.

Another point would be if the application should close or minimize into the taskbar?

torbei commented:
> From a logical perspective I think it shouldn't matter how the file is opened, the outcome should be the same.

I agree to that and I tried to do that, but couldn't get it to work. But I will look further into it.
Making it configurable is a great idea.

>> Another point would be if the application should close or minimize into the taskbar?
>
> This also could be made configurable.

Good idea. I will try it.

cboxdoerfer commented:
>I agree to that and I tried to do that, but couldn't get it to work. But I will look further into it.

Double click is handled here: https://github.com/cboxdoerfer/fsearch/blob/master/src/fsearch_window.c#L629

torbei commented:
Thanks, somehow I overlooked that.

torbei commented:
Should work now, if you can test it.
You can adjust if you want to minimize or close it.
As well as doing it after double-clicking or shortcut use.

I wanted to align the ComboBox left to the label but everything I tried didn't work. So until then it's align right if this is ok.

cboxdoerfer commented on src/fsearch_window_actions.c:
```diff
@@ -321,6 +321,20 @@ open_with_cb (GtkTreeModel *model,
+        if (config->action_after_file_open == 1) {
+            g_application_quit (G_APPLICATION (FSEARCH_APPLICATION_DEFAULT));
+        } else {
+            gtk_window_iconify (gtk_application_get_active_window (G_APPLICATION (FSEARCH_APPLICATION_DEFAULT)));
```
`gtk_application_get_active_window` expects a `GtkApplication *` so `G_APPLICATION` should rather be `GTK_APPLICATION`.

cboxdoerfer commented on src/fsearch_window_actions.c:
```diff
@@ -384,6 +398,7 @@ fsearch_window_action_open (GSimpleAction *action,
             gtk_tree_selection_selected_foreach (selection, open_cb, NULL);
         }
     }
+    fsearch_window_action_after_file_open(false);
```
This action should only be called when the file actually was opened. Right now the application even gets quit when no attempt to open a  file occurred, when more than 10 files were selected (yes, this arbitrary limit sucks, I'll need to add a configurable confirm dialog which at least warns the user when he tries to open lots of files). So I guess we should add some error handling here and only trigger the quit/minimize action when at least one file was successfully opened.

cboxdoerfer commented on src/fsearch_window_actions.c:
```diff
@@ -414,6 +429,7 @@ fsearch_window_action_open_folder (GSimpleAction *action,
             gtk_tree_selection_selected_foreach (selection, open_folder_cb, NULL);
         }
     }
+    fsearch_window_action_after_file_open(false);
```
Same issue as above.

cboxdoerfer commented on src/fsearch_window.c:
```diff
@@ -629,6 +629,7 @@ on_listview_row_activated (GtkTreeView       *tree_view,
             launch_node (node);
         }
     }
+    fsearch_window_action_after_file_open(true);
```
Same issue as below.

cboxdoerfer commented:
Patch looks good except those two issues. Do you want me to add the error handling? Then you just need to update the pull request to handle the new error codes?

torbei commented:
> Patch looks good except those two issues. Do you want me to add the error handling?

If you have time and motivation, you can totally do it.

> Oh and what do you think should happen when e.g. 3 files were selected but only 2 could be opened? 

Yes I think a warn dialog and remaining open would be better. Maybe with an option "Don't bother me again". Then you can ask what the user wants (Stay open / Minimize nevertheless)

cboxdoerfer commented:
@torbei, `launch_node` and `launch_node_path` now return `true`/`false` and I already the placeholder code in which you can place the close/minimize actions:
https://github.com/cboxdoerfer/fsearch/blob/master/src/fsearch_window_actions.c#L390
https://github.com/cboxdoerfer/fsearch/blob/master/src/fsearch_window_actions.c#L427
https://github.com/cboxdoerfer/fsearch/blob/master/src/fsearch_window.c#L630

torbei commented:
Great I will look at it tomorrow.

By the way what do you think about the case if opening fails?
Just a warning dialog or give the user the option what he wants?

cboxdoerfer commented:
@torbei, yes a warning dialog, which can be disabled in the preferences sounds good to me. It probably makes sense to add a whole section in the preferences to control all warning dialogs. Something like:

```
Show warning dialog when:
    [ ] file/folder failed to open
    [ ] attempting to delete a file/folder
    [x] attempting to delete many files/folders
    ...
```

But that's nothing you need to worry about for this pull request.

torbei commented:
@cboxdoerfer So I have a question. Did you test the return values of `launch_node`. If so, how did you do it?

I tried to create a file, update the database and then delete it from my system. So I have the entry in fsearch, but the file doesn't exist anymore.

Now if I want to open it, it will show the following error dialog:

> Unable to run the command specified. The file or folder /home/... does not exist. [OK]

But it won't execute the else section of the `if (!open_failed)`.

Do you have any idea how to test it?

cboxdoerfer commented:
@torbei, yes it only returned `false` when launching `xdg-open` fails. But I just added a new test which returns false when the file doesn't exist.

torbei commented:
Ok I think I have everything now.

So I implemented the dialog if the opening fails and ask what the user wants.
I wanted to include the "don't show again" option but I could find any support for it in the GTKDialog class.
So I added a new tab in the preference window where you can enable / disable the warning dialogs and added the dialog there.
If you disable it there, the application will ask for the default action of it and saves it.

I'm pretty happy with this solution but I don't know about you as the maintainer :)

cboxdoerfer commented:
@torbei, some issues I see with the UI:

1.  I don't like it when using the `file / folder failed to open` toggle button in the preferences pops up a new dialog window basically out of nowhere. If this toggle button leads to further options those options should be visible right away . Like that:

```
[x] Option 1
    [x] Option 1.1 
    [ ] Option 1.2
[ ] Option 2
    [ ] Option 2.1
```

2. And there's a more important issue with that dialog. The default action after opening a file/folder is to do nothing - i.e. leave the window visible. But when a user uses `file / folder failed to open` toggle button to disable just the warning message, he gets asked the following: 

> Do you want to keep the window open instead of closing it?

Which is not only confusing, because a user might wonder if this somehow overrides the other setting in the `Interface` tab, but also doesn't make any sense. The window isn't going to be closed, so there's no reason to ask that kind of question right now.

Solution: I don't think the dialog tab is the right place to ask for what's about to happen when opening a file/folder failed. This section is only supposed to configure the behaviour of dialogs. If you want to give the user the choice to configure both outcomes of the `open file/folder action` (1. what's supposed to happen when it succeeds or 2. fails) then those two outcomes should be configurable in the same place. However the UI in the `Interface` tab becomes much more complex then and I'm not sure if that's even worth it. Imho if file/folder open fails the minimize/quite action just shouldn't be activated.

torbei commented:
Yeah with a second thought I agree. UX wise it wasn't a good choice. I will change it.

> Imho if file/folder open fails the minimize/quit action just shouldn't be activated.

Ok but still throwing an dialog (which can be disabled) ?

cboxdoerfer commented:
>Ok but still throwing a dialog (which can be disabled) ?

Yes. 

Maybe to avoid some confusion we could change the checkbox label `Behaviour after opening a file` to `Behaviour after successfully opening a file`. Then it's clear that this action won't be triggered if file/folder opening fails.

torbei commented:
Ok I removed it.

cboxdoerfer commented:
Yes. Looks good now.

cboxdoerfer commented on src/fsearch_window.c:
```diff
@@ -628,6 +629,20 @@ on_listview_row_activated (GtkTreeView       *tree_view,
+            } else {
+                // open failed
+                FsearchConfig *config = fsearch_application_get_config (FSEARCH_APPLICATION_DEFAULT);
+                if (config->show_dialog_failed_opening) {
```
There's a bug here. This dialog should only be shown when `config->action_after_file_open_keyboard` or
`config->action_after_file_open_mouse` is `true`.

torbei commented:
Fixed it

cboxdoerfer commented:
Great. Merging :)
